### PR TITLE
Feature/full folder uid handling

### PR DESCRIPTION
**Pull request by axelsiebert**
## Changes

- Change all methods from Controller down to Access layer to use versioned_object_ids
- Change version_at_time to use Java Instants in UTC
- Change all Create and Update methods to use UTC timestamps
- Provide Etag header as expected

## Related issue

Closes https://github.com/ehrbase/project_management/issues/148

## Additional information and checks

- [x] Pull request linked in changelog

### CODE REVIEW

File: application/src/main/java/org/ehrbase/application/EhrBase.java [hunk 1]
```diff
@@ -23,9 +23,10 @@
 import org.ehrbase.rest.openehr.RestOpenehrModuleConfiguration;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.security.servlet.SecurityAutoConfiguration;
 import org.springframework.context.annotation.Import;
 
-@SpringBootApplication
+@SpringBootApplication(exclude = SecurityAutoConfiguration.class)
```
Comment (axelsiebert) [inline application/src/main/java/org/ehrbase/application/EhrBase.java:29]: Remove this exclusion

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/OpenehrDirectoryController.java [hunk 2]
```diff
@@ -303,7 +352,7 @@ public OpenehrDirectoryController(FolderService folderService, EhrService ehrSer
             @ApiParam(value = REQ_CONTENT_TYPE_BODY) @RequestHeader(value = CONTENT_TYPE) String contentType,
             @ApiParam(value = REQ_ACCEPT) @RequestHeader(value = ACCEPT, required = false, defaultValue = MediaType.APPLICATION_JSON_VALUE) String accept,
             @ApiParam(value = REQ_PREFER) @RequestHeader(value = PREFER, required = false, defaultValue = RETURN_MINIMAL) String prefer,
-            @ApiParam(value = "{preceding_version_uid}", required = true) @RequestHeader(value = IF_MATCH) String ifMatch,
+            @ApiParam(value = "{preceding_version_uid}", required = true) @RequestHeader(value = IF_MATCH) ObjectVersionId folderId,
             @ApiParam(value = "EHR identifier from resource path after ehr/", required = true) @PathVariable(value = "ehr_id") UUID ehrId,
             @ApiParam(value = "Update data for the target FOLDER") @RequestBody Folder folder,
             @RequestUrl String requestUrl
```
Comment (axelsiebert) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/OpenehrDirectoryController.java:358]: Remove unused custom annotation

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/OpenehrDirectoryController.java [hunk 3]
```diff
@@ -455,8 +558,33 @@ private HttpStatus getSuccessStatus(HttpMethod method) {
      * @return String is a valid path value or not
      */
     private boolean isValidPath(String path) {
-        Pattern pathPattern = Pattern.compile("^(?:/?(?:\\w+|\\s)*/?)+$");
+        Pattern pathPattern = Pattern.compile("^(?:/?(?:\\w+|\\s|-)*/?)+$");
         return pathPattern.matcher(path).matches();
     }
+
```
Comment (axelsiebert) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/OpenehrDirectoryController.java:564]: Add TODO comments for Archie fix

File: service/pom.xml [hunk 4]
```diff
@@ -136,6 +136,14 @@
             <artifactId>test-data</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.ehrbase.openehr</groupId>
+            <artifactId>rest-openehr</artifactId>
+        </dependency>
```
Comment (axelsiebert) [inline service/pom.xml:142]: Remove duplicate

File: service/src/main/java/org/ehrbase/dao/access/jooq/FolderHistoryAccess.java [hunk 5]
```diff
@@ -614,12 +621,12 @@ public static int getVersionFromTimeStamp(I_DomainAccess domainAccess, UUID vFol
                                 filteredHierarchicalTable.field("parent_folder", UUID.class).eq(folderUid))).asTable();
 
 
-        Result<Record> folderSelectedRecordSub = domainAccess.getContext().withRecursive("subfolders").as(
```
Comment (axelsiebert) [inline service/src/main/java/org/ehrbase/dao/access/jooq/FolderHistoryAccess.java:617]: Remove debug intermediate variable

File: service/src/main/java/org/ehrbase/dao/access/util/FolderUtils.java [hunk 6]
```diff
@@ -96,6 +98,32 @@ public static void checkSiblingNameConflicts(Folder folder) {
         }
     }
```
Comment (axelsiebert) [inline service/src/main/java/org/ehrbase/dao/access/util/FolderUtils.java:100]: Add Todo comment after archie fix

File: service/src/main/java/org/ehrbase/service/FolderServiceImp.java [hunk 7]
```diff
@@ -100,98 +96,124 @@ public UUID create(UUID ehrId, Folder content) {
...
-            return Optional.empty();
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Optional<FolderDto> getLatest(ObjectVersionId folderId, String path) {
+
+        I_FolderAccess folderAccess = I_FolderAccess.getInstanceForExistingFolder(
+                getDataAccess(),
+                folderId,
```
Comment (axelsiebert) [inline service/src/main/java/org/ehrbase/service/FolderServiceImp.java:177]: Change to use Instant for timestamp creation

File: service/src/main/java/org/ehrbase/service/FolderServiceImp.java [hunk 8]
```diff
@@ -201,15 +223,15 @@ public UUID create(UUID ehrId, Folder content) {
                                     FolderAccess.buildNewFolderAccessHierarchy(
                                             getDataAccess(),
                                             childFolder,
-                                            timestamp,
```
Comment (axelsiebert) [inline service/src/main/java/org/ehrbase/service/FolderServiceImp.java:204]: Change to use Instant for timestamp

File: service/src/main/java/org/ehrbase/service/FolderServiceImp.java [hunk 9]
```diff
@@ -308,15 +343,27 @@ public Integer getVersionNumberForTimestamp(
...
     private Folder createFolderObject(I_FolderAccess folderAccess) {
 
-        Folder result = new Folder();
-        result.setDetails(folderAccess.getFolderDetails());
-        result.setArchetypeNodeId(folderAccess.getFolderArchetypeNodeId());
-        result.setNameAsString(folderAccess.getFolderName());
-        result.setItems(folderAccess.getItems());
-        result.setUid(new ObjectVersionId(folderAccess.getFolderId()
-                .toString()));
+        Folder result = new Folder(
```
Comment (axelsiebert) [inline service/src/main/java/org/ehrbase/service/FolderServiceImp.java:346]: Check for null values in constructor

**Review by jakesmolka — APPROVED** [review]

**Pull request merged**